github/release_notes/renderer.py

```diff
@@ -0,0 +1,280 @@
+        nodes=None,
+        matches_rn_field_path='target_group_id'
+    ), \
+    Node(
+        identifier='operator',
+        title='OPERATOR',
+        nodes=None,
+        matches_rn_field_path='target_group_id'
+    )
+categories = \
+    Node(
+        identifier='noteworthy',
+        title='Most notable changes',
+        nodes=target_groups,
+        matches_rn_field_path='category_id'
+    ), \
+    Node(
+        identifier='improvement',
+        title='Improvements',
+        nodes=target_groups,
+        matches_rn_field_path='category_id'
+    )
+
+class Renderer(object):
+
+    def __init__(self, release_note_objs: list):
+        self.release_note_objs = release_note_objs
+
+    def render(self)->str:
+        origin_nodes = _\
+            .chain(self.release_note_objs)\
+            .sort_by(lambda rn_obj: rn_obj.cn_source_repo.github_repo())\
+            .uniq_by(lambda rn_obj: rn_obj.cn_source_repo.name())\
+            .map(lambda rn_obj: Node(
+                identifier=rn_obj.cn_source_repo.name(),
+                title='[{origin_name}]'.format(origin_name=rn_obj.cn_source_repo.github_repo()),
+                nodes=categories,
+                matches_rn_field_path='cn_source_repo.name' # path points to a function
+            ))\
+            .value()
+
+        md_lines = self._nodes_to_lines(
+            nodes=origin_nodes,
+            level=1,
+            release_note_objs=self.release_note_objs
+        )
+
+        if md_lines:
+            return '\n'.join(md_lines)
+        else: # fallback
+            return 'no release notes available'
+
+    def _nodes_to_lines(
+        self,
+        nodes: list,
+        level: int,
+        release_note_objs: list
+    )->list:
+        lines = list()
+        for node in nodes:
+            filtered_rn_objects = _.filter(
+                release_note_objs,
+                lambda rn: node.identifier == get_or_call(rn, node.matches_rn_field_path)
+            )
+            if not filtered_rn_objects:
+                continue
+            if node.nodes:
+                tmp_lines = self._nodes_to_lines(
+                    nodes=node.nodes,
+                    level=level + 1,
+                    release_note_objs=filtered_rn_objects
+                )
+                skip_title = False
+            else:
+                tmp_lines = self._to_bullet_points(
+                    tag=node.title,
+                    rn_objs=filtered_rn_objects
+                )
+                # title is used as bullet point tag -> no need for additional title
+                skip_title = True
+
+            # only add title if there are lines below the title
+            if tmp_lines:
+                if not skip_title:
+                    lines.append(self._title(node, level))
+                lines.extend(tmp_lines)
+        return lines
+
+    def _header_suffix(
+        self,
+        rn_obj: ReleaseNote
+    )->str:
+        header_suffix = ''
```

Comment:
this will lead to 
```
return header_suffix
UnboundLocalError: local variable 'header_suffix' referenced before assignment
```
and same reasoning as above that there is an empty string returned in case there is no user_login and reference_id

Reply:
can't you invert it in this case and do a check + early exit in this case?